csrc/scheduler/mma_utils.cpp

```diff
@@ -1437,103 +1368,195 @@ TensorView* getTensorviewPriorToCast(TensorView* in) {
-          // If the Mul-Sum is a valid representation of a matmul,
-          // then get the properties of the replacement Mma op.
-          auto props = getMulSumInsOutsBcasts(bOp, stmt);
-          if (props.has_value()) {
-            mul_sum_props_.push_back(*props);
+namespace {
+
+class MatmulPatternMatcher : IterVisitor {
+ public:
+  static std::vector<MatmulPattern> run(Fusion* fusion) {
+    MatmulPatternMatcher matcher;
+    matcher.traverse(fusion);
+    return matcher.patterns_;
+  }
+
+ private:
+  using IterVisitor::handle;
+
+  // Handle the case when no translation is needed.
+  void handle(MmaOp* mop) override {
+    MatmulPattern& pattern = patterns_.emplace_back();
+    pattern.A = mop->inA()->as<TensorView>();
+    pattern.B = mop->inB()->as<TensorView>();
+    pattern.output = mop->out()->as<TensorView>();
+  }
+
+  void handle(ReductionOp* rop) override {
+    // Check if operation is a sum.
+    if (rop->getReductionOpType() != BinaryOpType::Add) {
+      return;
+    }
+    // Then check if the producer of the sum is a mul.
+    if (auto bop = dynamic_cast<BinaryOp*>(rop->in()->definition())) {
+      if (bop->getBinaryOpType() != BinaryOpType::Mul) {
+        return;
+      }
+      // Remember that we are just gathering the immediate inputs to the
+      // matmul, so there should be no prologue between a, b and the mul/sum.
+
+      // Check that the inputs have broadcasts that are not all in common, i.e.
+      // that there is at least one M and at least one N dimension.
+
+      // Note that there might be a cast to Float just before the multiply. This
+      // happens when using the `mul` op with reduced precision inputs. It can
+      // also happen if the inputs to `mul` in the definition were Float, but
+      // the Fusion was segmented and casts to half precision were inserted at
+      // the segmentation edge (see castInputOutputToLowerPrecision in
+      // fusion_segmenter.cpp).
+      TensorView* ltv = getTensorviewPriorToCast(bop->lhs()->as<TensorView>());
+      TensorView* rtv = getTensorviewPriorToCast(bop->rhs()->as<TensorView>());
+
+      std::vector<IterDomain*> lrf =
+          TensorDomain::noReductions(ltv->getMaybeRFactorDomain());
+      std::vector<IterDomain*> rrf =
+          TensorDomain::noReductions(rtv->getMaybeRFactorDomain());
+
+      // These sizes should match since ops::maybeBroadcast places BroadcastOps
+      // for implicit broadcasting.
+      NVF_ERROR(lrf.size() == rrf.size());
+      const std::vector<IterDomain*>& red_root =
+          rop->out()->as<TensorView>()->getRootDomain();
+      NVF_ERROR(red_root.size() == lrf.size());
+      bool has_m = false, has_n = false;
+      for (size_t i : c10::irange(lrf.size())) {
+        if (lrf[i]->isBroadcast() && !rrf[i]->isBroadcast()) {
+          has_m = true;
+        } else if (!lrf[i]->isBroadcast() && rrf[i]->isBroadcast()) {
+          has_n = true;
+        }
```

Comment:
Looks like this is not worse that what is already in main, but I don't think it makes sense to assume lhs is A and rhs is B. It can be the opposite. For now, can we assume the output's rfactor domain is in the order of `m,n,k`, and use this to infer which is A and which is B here? We can review this piece of code when we support matmul-transpose fusion so that we do not assume output order `m,n,k`.

Reply:
I agree. Ideally we would use the output allocation domain and IdModel to determine which is A and which is B for this pattern. When I wrote this pass I was not using those yet, but I could plumb them in. Since this takes place before translation, then we would need to rebuild the IdModel after translation.

For now, as you said I will assume the rfactor is the allocation domain for determining A and B, and I'll place a note there that this should be revisited.

Reply:
I just pushed a change that looks at the reduction output root domain to determine N which lets us determine which input is B and which is A.